### PR TITLE
[Notifier] [SpotHit] Add the bridge

### PR DESCRIPTION
**Pull request by JamesHemery**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | symfony/symfony-docs#14869
| Recipe PR        | symfony/recipes#881

@OskarStark Bridge added :)

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/SpotHit/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+Spot-Hit Notifier
+====================
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/README.md]: This underline is too long

File: src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+use Symfony\Contracts\HttpClient\Exception\ServerExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author James Hemery <james@yieldstudio.fr>
+ */
+final class SpotHitTransport extends AbstractTransport
+{
+    protected const HOST = 'www.spot-hit.fr/api/envoyer/sms';
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php]:
```suggestion
    protected const HOST = 'spot-hit.fr';
```
The "/api/...." part should only be added to the POST down below.

File: src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    private $token;
+    private $from;
+
+    public function __construct(
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php]: Please construct on one line

File: src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransportFactory.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        $scheme = $dsn->getScheme();
+        $token = $this->getUser($dsn);
+        $from = $dsn->getOption('from');
+        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
+
+        if ('spothit' === $scheme) {
+            return (new SpotHitTransport($token, $from, $this->client, $this->dispatcher))->setHost($host);
+        }
+
+        throw new UnsupportedSchemeException($dsn, 'spothit', $this->getSupportedSchemes());
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransportFactory.php]:
```suggestion
        $scheme = $dsn->getScheme();

        if ($scheme !== 'spothit) {
            throw new UnsupportedSchemeException($dsn, 'spothit', $this->getSupportedSchemes());
        }

        $token = $this->getUser($dsn);
        $from = $dsn->getOption('from');
        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
        $port = $dsn->getPort();

        return (new SpotHitTransport($token, $from, $this->client, $this->dispatcher))->setHost($host)->setPort($port);       
```

File: src/Symfony/Component/Notifier/Bridge/SpotHit/composer.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{
+    "name": "symfony/spot-hit-notifier",
+    "type": "symfony-bridge",
+    "description": "Symfony SpotHit Notifier Bridge",
+    "keywords": ["sms", "Spot-Hit", "notifier", "symfony"],
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/composer.json]:
```suggestion
    "keywords": ["sms", "spot-hit", "notifier", "symfony"],
```

File: src/Symfony/Component/Notifier/Bridge/SpotHit/phpunit.xml.dist [hunk 6]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/phpunit.xml.dist:1]: Can you please copy this file from an existing bridge to keep the same formatting? Thanks

File: src/Symfony/Component/Notifier/Bridge/SpotHit/README.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
+Spot-Hit Notifier
+=================
+
+Provides Spot-Hit integration for Symfony Notifier.
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/README.md]:
```suggestion
Provides [Spot-Hit](https://....) integration for Symfony Notifier.
```

Please add a URL

File: src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+ * @author James Hemery <james@yieldstudio.fr>
+ */
+final class SpotHitTransport extends AbstractTransport
+{
+    protected const HOST = 'spot-hit.fr';
+
+    private $token;
+    private $from;
+
+    public function __construct(string $token, ?string $from = null, HttpClientInterface $client = null, EventDispatcherInterface $dispatcher = null) {
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php]:
```suggestion
    public function __construct(string $token, string $from = null, HttpClientInterface $client = null, EventDispatcherInterface $dispatcher = null)
    {
```

File: src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            throw new UnsupportedMessageTypeException(__CLASS__, SmsMessage::class, $message);
+        }
+
+        $response = $this->client->request('POST', 'https://www.'.$this->getEndpoint().'/api/envoyer/sms', [
+            'body' => [
+                'key' => $this->token,
+                'destinataires' => $message->getPhone(),
+                'type' => 'premium',
+                'message' => $message->getSubject(),
+                'expediteur' => $this->from,
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php:80]: Is null here supported?

File: src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+     * @throws RedirectionExceptionInterface
+     * @throws ServerExceptionInterface
+     */
+    protected function doSend(MessageInterface $message): SentMessage
+    {
+        if (!$this->supports($message)) {
+            throw new UnsupportedMessageTypeException(__CLASS__, SmsMessage::class, $message);
+        }
+
+        $response = $this->client->request('POST', 'https://www.'.$this->getEndpoint().'/api/envoyer/sms', [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php]:
```suggestion
        $endpoint = sprintf('https://www.%s/api/envoyer/sms', $this->getEndpoint());
        $response = $this->client->request('POST', $endpoint, [
```

and do we really need the www. part?

File: src/Symfony/Component/Notifier/Bridge/SpotHit/composer.json [hunk 11]
```diff
@@ -0,0 +1,35 @@
+{
+    "name": "symfony/spot-hit-notifier",
+    "type": "symfony-bridge",
+    "description": "Symfony SpotHit Notifier Bridge",
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/composer.json]:
```suggestion
    "description": "Symfony Spot-Hit Notifier Bridge",
```

File: src/Symfony/Component/Notifier/Bridge/SpotHit/composer.json [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/http-client": "^4.3|^5.1",
+        "symfony/notifier": "^5.3",
+        "symfony/event-dispatcher-contracts": "^2"
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/composer.json]:
```suggestion
```

**Review by OskarStark — COMMENTED** [review]
Just some minor last comments

File: src/Symfony/Component/Notifier/Bridge/SpotHit/Tests/SpotHitTransportFactoryTest.php [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+namespace Symfony\Component\Notifier\Bridge\SpotHit\Tests;
+
+use Symfony\Component\Notifier\Bridge\SpotHit\SpotHitTransportFactory;
+use Symfony\Component\Notifier\Tests\TransportFactoryTestCase;
+use Symfony\Component\Notifier\Transport\TransportFactoryInterface;
+
+final class SpotHitTransportFactoryTest extends TransportFactoryTestCase
+{
+    /**
+     * @return SpotHitTransportFactory()
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/Tests/SpotHitTransportFactoryTest.php]:
```suggestion
     * @return SpotHitTransportFactory
```

File: src/Symfony/Component/Notifier/Bridge/SpotHit/Tests/SpotHitTransportTest.php [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Tests\TransportTestCase;
+use Symfony\Component\Notifier\Transport\TransportInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+final class SpotHitTransportTest extends TransportTestCase
+{
+    /**
+     * @return SpotHitTransport()
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/Tests/SpotHitTransportTest.php]:
```suggestion
     * @return SpotHitTransport
```

File: src/Symfony/Component/Notifier/Bridge/SpotHit/composer.json [hunk 15]
```diff
@@ -0,0 +1,35 @@
+{
+    "name": "symfony/spot-hit-notifier",
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/SpotHit/composer.json]:
```suggestion
    "name": "symfony/spothit-notifier",
```

File: src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php [hunk 16]
```diff
@@ -68,6 +68,10 @@ class UnsupportedSchemeException extends LogicException
             'class' => Bridge\FreeMobile\FreeMobileTransportFactory::class,
             'package' => 'symfony/free-mobile-notifier',
         ],
+        'spothit' => [
+            'class' => Bridge\SpotHit\SpotHitTransportFactory::class,
+            'package' => 'symfony/spot-hit-notifier',
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php]:
```suggestion
            'package' => 'symfony/spothit-notifier',
```

**Comment by OskarStark** [comment]
@JamesHemery open to finish this PR? 😃

File: src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php (hunk 9 above)
Comment (JamesHemery) [reply src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php:80]: Yes is optionnal parameter

File: src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php (hunk 10 above)
Comment (JamesHemery) [reply src/Symfony/Component/Notifier/Bridge/SpotHit/SpotHitTransport.php]: It's specified in the Spot-Hit documentation.

**Comment by OskarStark** [comment]
One last comment, lets name name package `symfony/spothit-notifer` please, same for docs and recipe, afterwards good to go 👍

**Review by OskarStark — APPROVED** [review]

**Review by derrabus — APPROVED** [review]

**Comment by OskarStark** [comment]
Thanks James for working on this feature, this is much appreciated.

**Comment by fabpot** [comment]
FYI, I've fixed the package name in #40186

**Pull request merged**